From 17036976bbec57f7d616156794501838cdaac421 Mon Sep 17 00:00:00 2001
From: Quentin Casasnovas <quentin.casasnovas@vates.tech>
Date: Mon, 5 Jan 2026 20:23:35 +0100
Subject: [PATCH] THIS PATCH IS NOT INTENDED TO BE UPSTREAMED, IT HAS ONLY
 INFORMING PURPOSES!

I've noticed a performance regression with upstream kernels when used as Dom0
under Xen. The classic kernel can utilize the whole bandwidth of a 10G NIC
(ca. 9.3 Gbps), but upstream can reach only ca. 7 Gbps. I found that it
happens because SWIOTLB has to do double buffering. The per task frag
allocator introduced in 5640f7 creates 32 kb frags, which are not contiguous
in mfn space.
This patch provides a workaround by going back to the old way. The possible
ideas came up to solve this:

* make sure Dom0 memory is contiguous: it sounds trivial, but doesn't work with
driver domains, and there are lots of situations where this is not possible.
* use PVH Dom0: so we will have IOMMU. In the future sometime.
* use IOMMU with PV Dom0: this seems to happen earlier.

Signed-off-by: Zoltan Kiss <zoltan.kiss@citrix.com>
Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>

[Quentin: rebase from v4.19.19 to v4.19.325 conflicts:
 - net/core/sock.c: SKB_FRAG_PAGE_ORDER moved to include/net/sock.h in
   ce8908796665 ("esp: Fix possible buffer overflow in ESP transformation")]

Signed-off-by: Quentin Casasnovas <quentin.casasnovas@vates.tech>
---
 include/net/sock.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/net/sock.h b/include/net/sock.h
index 507ed9e5cf3c..77335c541e92 100644
--- a/include/net/sock.h
+++ b/include/net/sock.h
@@ -2595,7 +2595,7 @@ extern __u32 sysctl_wmem_default;
 extern __u32 sysctl_rmem_default;
 
 /* On 32bit arches, an skb frag is limited to 2^15 */
-#define SKB_FRAG_PAGE_ORDER	get_order(32768)
+#define SKB_FRAG_PAGE_ORDER	get_order(4096)
 
 static inline int sk_get_wmem0(const struct sock *sk, const struct proto *proto)
 {
-- 
2.43.0

